keps/sig-network/0013-20180612-ipv4-ipv6-dual-stack.md

```diff
@@ -0,0 +1,626 @@
+- The pod status API changes will include a per-IP string map for arbitrary annotations, as a placeholder for future Kubernetes enhancements. This mapping is not required for this dual-stack design, but will allow future annotations, e.g. allowing a CNI network plugin to indicate to which network a given IP address applies.
+- Kubectl commands and output displays will need to be modified for dual-stack.
+- Kubeadm support will need to be added to enable spin-up of dual-stack clusters. Kubeadm support is required for implementing dual-stack continuous integration (CI) tests.
+- New e2e test cases will need to be added to test parallel IPv4/IPv6 connectivity between pods, nodes, and services.
+
+### Awareness of Multiple IPs per Pod
+
+Since Kubernetes Version 1.9, Kubernetes users have had the capability to use dual-stack-capable CNI network plugins (e.g. Bridge + Host Local, Calico, etc.), using the 
+[0.3.1 version of the CNI Networking Plugin API](https://github.com/containernetworking/cni/blob/spec-v0.3.1/SPEC.md), to configure multiple IPv4/IPv6 addresses on pods. However, Kubernetes currently captures and uses only IP address from the pod's main interface.
+
+This proposal aims to extend the Kubernetes Pod Status API so that Kubernetes can track and make use of up to one IPv4 address and up to one IPv6 address assignment per pod.
+
+#### Versioned API Change: PodStatus v1 core
+In order to maintain backwards compatibility for the core V1 API, this proposal retains the existing (singular) "PodIP" field in the core V1 version of the [PodStatus V1 core API](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.10/#podstatus-v1-core), and adds a new array of structures that store pod IPs along with associated metadata for that IP. The metadata for each IP (refer to the "Properties" map below) will not be used by the dual-stack feature, but is added as a placeholder for future enhancements, e.g. to allow CNI network plugins to indicate to which physical network that an IP is associated. Retaining the existing "PodIP" field for backwards compatibility is in accordance with the [Kubernetes API change quidelines](https://github.com/kubernetes/community/blob/master/contributors/devel/api_changes.md).
+```
+    // Default IP address allocated to the pod. Routable at least within the
+    // cluster. Empty if not yet allocated.
+    PodIP string `json:"podIP,omitempty" protobuf:"bytes,6,opt,name=podIP"`
+
+    // IP address information for entries in the (plural) PodIPs slice.
+    // Each entry includes:
+    //    IP: An IP address allocated to the pod. Routable at least within
+    //        the cluster.
+    //    Properties: Arbitrary metadata associated with the allocated IP.
+    type PodIPInfo struct {
+        IP string
+        Properties map[string]string
```

Comment:
The properties map would be very useful for multi-network. And no sense in changing the data structure twice? I'd prefer to keep it if possible.

Reply:
I see this has already generated some discussion :)
I'll add my own comments here - @leblancd you can ignore my other comment up above.

One pattern I've seen used successfully in internal interfaces is to have a mostly-strongly typed struct with a bag-of-strings at the end for "experimental" free-for-all properties. But that requires agreement that the bag-of-strings should not be relied upon, and can change at any time. I do not think we can enforce such a thing if we put this into an external facing API, so my vote is to only add fully-typed fields with validation and strong semantic meaning. Then we can argue about names all at once before they are used instead of after the fact :)